cluster.go

```diff
@@ -262,6 +291,79 @@ func (c *clusterClient) _refresh() (err error) {
 			}
 		}(removes)
 	}
+}
+
+func (c *clusterClient) conditionalRefresh() (err error) {
+	result, err := c.getClusterTopology()
+	if err != nil {
+		return err
+	}
+
+	groups := result.parse(c.opt.TLSConfig != nil)
+
+	// we need to check whether the new topology is different from the current one.
+	// so we don't need to early re-create the connections.
+	conns := make(map[string]connrole, len(groups))
+	for master, g := range groups {
+		conns[master] = connrole{replica: false}
+		for _, addr := range g.nodes[1:] {
+			if c.rOpt != nil {
+				conns[addr] = connrole{replica: true}
+			} else {
+				conns[addr] = connrole{replica: true}
+			}
+		}
+	}
+	// make sure InitAddress always be present
+	for _, addr := range c.opt.InitAddress {
+		if _, ok := conns[addr]; !ok {
+			conns[addr] = connrole{}
+		}
+	}
+
+	isChanged := false
+	c.mu.RLock()
+	// check if the new topology is different from the current one
+	for addr, cc := range conns {
+		old, ok := c.conns[addr]
+		if !ok || old.replica != cc.replica {
+			isChanged = true
+			break
+		}
+	}
+	// check if the current topology is different from the new one
+	if !isChanged {
+		for addr := range c.conns {
+			if _, ok := conns[addr]; !ok {
+				isChanged = true
+				break
+			}
+		}
+	}
+	c.mu.RUnlock()
+
+	if !isChanged {
+		return nil
+	}
+
+	for addr, cc := range conns {
+		if cc.replica {
+			if c.rOpt != nil {
+				cc.conn = c.connFn(addr, c.rOpt)
+			} else {
+				cc.conn = c.connFn(addr, c.opt)
+			}
+		} else {
+			cc.conn = c.connFn(addr, c.opt)
+		}
+
+		conns[addr] = connrole{
+			conn:    cc.conn,
+			replica: cc.replica,
+		}
+	}
```

Comment:
This loop costs nothing because `connFn` doesn't connect to the address immediately. So, you don't need to separate it from the loop above.